messaging/src/main/java/org/axonframework/configuration/RootConfigurer.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.configuration;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.concurrent.TimeUnit;
+
+import jakarta.annotation.Nonnull;
+
+/**
+ * The root {@link NewConfigurer configurer} of any Axon Framework application.
+ * <p>
+ * Provides a means to register {@link #onStart(int, LifecycleHandler) start} and
+ * {@link #onShutdown(int, LifecycleHandler) shutdown} handlers for this application, besides containing all
+ * {@link Component components}, {@link ComponentDecorator component decorators}, and {@link Module modules}.
+ * <p>
+ * Once the configuring phase is completed, the {@link RootConfiguration} of the application can be retrieved by
+ * invoking {@link #build()}. Starting the {@code NewConfiguration} can be done through a separate
+ * {@link RootConfiguration#start()} invocation, or {@link #start()} can be used to build-and-start the project
+ * immediately.
+ *
+ * @author Allard Buijze
+ * @author Steven van Beelen
+ * @since 3.0.0
+ */
+public interface RootConfigurer extends StartableConfigurer<RootConfigurer> {
+
+    /**
+     * Returns a {@code RootConfigurer} instance to start configuring {@link Component components},
+     * {@link ComponentDecorator component decorators}, {@link ConfigurerEnhancer enhancers}, and {@link Module modules}
+     * for an Axon Framework application.
+     *
+     * @return A {@code RootConfigurer} instance for further configuring.
+     */
+    static RootConfigurer defaultConfigurer() {
+        return configurer(true);
+    }
+
+    /**
+     * Returns a {@code RootConfigurer} instance to start configuring {@link Component components},
+     * {@link ComponentDecorator component decorators}, {@link ConfigurerEnhancer enhancers}, and {@link Module modules}
+     * for an Axon Framework application.
+     * <p>
+     * When {@code autoLocateEnhancers} is {@code true}, a {@link ServiceLoader} will be used to locate all declared
+     * instances of type {@link ConfigurerEnhancer}. Each of the discovered instances will be invoked, allowing it to
+     * set default values for the returned {@code RootConfigurer}.
+     *
+     * @param autoLocateEnhancers Flag indicating whether {@link ConfigurerEnhancer} on the classpath should be
+     *                            automatically retrieved. Should be set to {@code false} when using an application
+     *                            container, such as Spring or CDI.
+     * @return A {@code RootConfigurer} instance for further configuring.
+     */
+    static RootConfigurer configurer(boolean autoLocateEnhancers) {
```

Comment:
I'm not sure if it's OK or should we provide interface there like `EnhancersLoader` and pass `NoOp` implementation in case of Spring and abstract away what we have here to `SerivceLoaderEnhancersLoader`, then we can avoid the if here.

Reply:
Can you explain to me why the if-statement is an issue here?

Not that an `EnhancerLoader` or something similar wouldn't work, but it seems like a lot of extra effort to remove a single if-statement.

Now, whether we should keep the `ServiceLoader` mechanism for this, that's a discussion I would like to have. :-)

Reply:
Just `if` it's not a problem a t all, but I think about the extendability - the class is public and the boolean limit it to just the usage of `ServiceLoader` or not at all. 
I can image someone may want to locate enhancers in a different way - I don't know in which way right now :D

Reply:
Hmm, interesting. Let's chat about it shortly in our AF5-meetings. We can always make this a follow-up issue. Perhaps having nicer extandability here will come in handy for (as you shared) Spring, or Micronaut, or Quarkus.